update msal-browser caching doc (#5047)
AzureAD/microsoft-authentication-library-for-js: lib/msal-browser/docs/caching.md

# Caching in MSAL

> When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. APIs like `acquireTokenSilent()` retrieves access tokens from the cache for a given account.

Reviewer: ```suggestion
When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. The `acquireTokenSilent()` API retrieves access tokens from the cache for a given account and renews them if needed.
```

| Cache Location   | Cleared on              | Shared between windows/tabs | Redirect flow supported |
|------------------|-------------------------|-----------------------------|-------------------------|
| `sessionStorage` | window/tab close        | No                          | Yes                     |
| `localStorage`   | user logout             | Yes                         | Yes                     |
| `memoryStorage`  | page refresh/navigation | No                          | No                      |

> :bulb: While the authentication state may be lost in session and memory storage due to window/tab close or page refresh/navigation, respectively, users will still have an active session with the IdP as long as the session cookie is not expired and might be able to re-authenticate without any prompts.

### Cookie storage

> MSAL Browser can be configured to use cookies for storing temporary authentication artifacts. This is useful for apps that need to support older browsers such as Internet Explorer. Note that when this option is chosen, tokens themselves are still stored in browser or memory storage. Please refer to [configuration](./configuration.md#cache-config-options) for more.

Reviewer 2: Isn't internet explorer officially end of life at this point?  What still supported browsers **require** this?  If we can suggest folks not use cookies that would be my preference.

Reviewer 3: I would rephrase this to specifically note that it is needed for browsers that may clear local/session storage during redirect-based login, which includes IE and Firefox Inprivate.

Reviewer 3: (as opposed to saying this is needed only for older browsers)

Author: revised!

### Security

> We consider session/local storage secure as long as your application does not have cross-site scripting (XSS) and related vulnurabilities. Please refer to the [OWASP XSS Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html) for securing your applications against XSS. If you are still concerned, we recommend using the `memoryStorage` option instead.

Reviewer 2: Do we need to flag the OWASP page as being an external link?

Author: Hmm, I would say orgs like OWASP and MDN need no special mentioning, but if we were to flag these, what would that look like? :/

## Cached artifacts

To faciliate efficient token acquisition while maintaining a good UX, MSAL caches various artifacts resulting from its API calls. At a glance, these are:

> - Persistent artifacts 

Reviewer 2: What does persistent mean here?  AT/IDToken might last for an hour, is that temporary?  How does that compare to the definition of Temporary on line 51 below.

Reviewer 3: Persistent would be anything that lasts for longer than the duration of the request (i.e. tokens, accounts)

Author: reworded this section and added a link to token lifetimes doc in case if there's confusion

# Caching in MSAL

> When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. The following MSAL APIs will perform one or more cache lookups to acquire tokens or retrieve accounts:

Reviewer: Technically, all of our APIs perform cache lookups but only `acquireTokenSilent` returns tokens from the cache. IMO this section should only mention `acquireTokenSilent`

Author: Ah, very well, reverted this

- `acquireTokenSilent()`
- `ssoSilent()`
- `getAccounts()`
- `getActiveAccount()`
- `getAccountByHomeId()`
- `getAccountByLocalId()`

## Cache storage

You can configure the cache storage location via the configuration object that is used to instantiate MSAL:

```typescript
import { PublicClientApplication, BrowserCacheLocation } from "@azure/msal-browser";

const pca = new PublicClientApplication({
    auth: {
        clientId: "Enter_the_Application_Id_Here", // e.g. "b1b60dca-c49d-496e-9851-xxxxxxxxxxxx" (guid)
        authority: "https://login.microsoftonline.com/Enter_the_Tenant_Info_Here", // e.g. "common" or your tenantId (guid),
        redirectUri: "/"
    },
    cache: {
       cacheLocation: BrowserCacheLocation.SessionStorage // "sessionStorage"
    }
});
```

> By default, MSAL stores the various authentication artifacts it obtains from the IdP in browser storage using the [Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API) supported by all modern browsers. Accordingly, MSAL offers two methods of storage: `sessionStorage` (default) and `localStorage`. In addition, MSAL provides `memoryStorage` option which allows you to opt-out of storing the cache in browser storage.

Reviewer: ```suggestion
By default, MSAL stores the various authentication artifacts it obtains from the IdP in browser storage using the [Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API) supported by all modern browsers. Accordingly, MSAL offers two methods of persistent storage: `sessionStorage` (default) and `localStorage`. In addition, MSAL provides `memoryStorage` option which allows you to opt-out of storing the cache in browser storage.
```

| Cache Location   | Cleared on              | Shared between windows/tabs | Redirect flow supported |
|------------------|-------------------------|-----------------------------|-------------------------|
| `sessionStorage` | window/tab close        | No                          | Yes                     |
| `localStorage`   | user logout             | Yes                         | Yes                     |
| `memoryStorage`  | page refresh/navigation | No                          | No                      |

> :bulb: While the authentication state may be lost in session and memory storage due to window/tab close or page refresh/navigation, respectively, users will still have an active session with the IdP as long as the session cookie is not expired and might be able to re-authenticate without any prompts.

### Cookie storage

> MSAL Browser can be configured to use cookies for storing the authentication state. This is useful for apps that need to support older browsers such as Internet Explorer. Note that when this option is choosed, tokens themselves are still stored in browser or memory storage. Please refer to [configuration](./configuration.md#cache-config-options) for more.

Reviewer: ```suggestion
MSAL Browser can be configured to use cookies for storing temporary authentication artifacts. This is useful for apps that need to support older browsers such as Internet Explorer. Note that when this option is chosen, tokens themselves are still stored in browser or memory storage. Please refer to [configuration](./configuration.md#cache-config-options) for more.
```

### Security

We consider session/local storage secure as long as your application does not have cross-site scripting (XSS) and related vulnurabilities. Please refer to the [OWASP XSS Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html) for securing your applications against XSS. If you are still concerned, we recommend using the `memoryStorage` option instead.

## Cached artifacts

To faciliate efficient token acquisition while maintaining a good UX, MSAL caches various artifacts resulting from its API calls. At a glance, these are:

- Persistent artifacts 
    - access tokens
    - id tokens
    - refresh tokens
    - accounts
- Temporary artifacts
    - request metadata (e.g. state, nonce, authority)
    - errors
    - interaction status
- Telemetry
    - previous failed request 
    - performance data

> :bulb: Temporary cache entries will always be stored in session storage or in memory. MSAL will fallback to memory storage if local/session storage is not available.

> :bulb: The authorization code is only stored in memory and will be discarded after redeeming it for tokens.

## Remarks

> - MSAL Browser's cache schema is compatible with other MSALs. However, there is no cache sharing support at the moment. 

Reviewer: This can be removed as it's not relevant at this time

Author: Sure thing, removed

# Caching

> When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. APIs like `acquireTokenSilent()` retrieves tokens from the cache for a given account.

Reviewer 4: Can we list all the APIs that retrieve cached tokens?

Author: Just added. Would that work?

```javascript
const msal = require('@azure/msal-browser');

// Create the msal application object
const pca = new msal.PublicClientApplication({
    auth: {
        clientId: "Enter_the_Application_Id_Here", // e.g. "b1b60dca-c49d-496e-9851-xxxxxxxxxxxx" (guid)
        authority: "https://login.microsoftonline.com/Enter_the_Tenant_Info_Here", // e.g. "common" or your tenantId (guid),
        redirectUri: "/"
    },
    cache: {

>        cacheLocation: "sessionStorage"

Reviewer 3: Can we show use the enum?

Author: done

    }
});

/**
* login* and acquireToken* APIs return an account object containing "homeAccountId"

> * you should keep a record of this in your app and use it later on when calling acquireTokenSilent

Reviewer 3: Instead, should mention using `setActiveAccount`.

Reviewer 3: Also, the code shippets below probably aren't needed for this doc.

Author: removed the snippet

* For more, see: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/accounts.md
*/
const someUserHomeAccountId = "Enter_User_Home_Account_Id";

const account = await pca.getAccountByHomeId(someUserHomeAccountId);

const silentTokenRequest = {
    account: account,
    scopes: ["user.read"],
};

pca.acquireTokenSilent(silentTokenRequest)
    .then((response) => {
        // do something with response
    }).catch((error) => {
        // catch and handle errors
    });
```

## Cache storage

By default, MSAL stores the various authentication artifacts it obtains from the IdP in browser storage using the [Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API) supported by all modern browsers. Accordingly, MSAL offers two methods of storage:

- `sessionStorage` (default):
- `localStorage`

> `sessionStorage` is more secure, but `localStorage` allows application or application instances on the same domain to share cache. This enables single sign-on between tabs and user sessions.

Reviewer 3: "more secure" is a bit of a misnomer. Instead, it should be noted that session storage is not shared between tabs and is cleared when the tab is closed. Probably best to have a table to compares localstorage/sessionstorage/memorystorage as to when the cache is cleared (persistent, on tab close, on page refresh/navigation, etc), whether it is shared between tabs, and whether redirect operations are supported (not supported with memory storage).

Reviewer 3: Also worth calling out that temporary cache entries will always be stored in session storage or in memory, and that msal will fallback to memory storage if local/session storage are not available.

Author: revised and added a table

Additionally, you can opt-out of browser storage by setting the `cacheLocation` to `memoryStorage` in msal configuration. This means that your application will lose authentication state upon page refresh (however, since users will still have an active session with the IdP, they would be able to re-authenticate without any prompts).

> We encourage you to explore the options and make the best decision for your application.

Reviewer 3: Can probably remove this sentence.

Author: done

### Cookie storage

> MSAL Browser can be configured to use cookies for storing the authentication state, in addition to browser or memory storage. This is useful for apps that need to support older browsers such as Internet Explorer. Please refer to [configuration](./configuration.md#cache-config-options) for more.

Reviewer 3: Should be made clear that tokens are not stored in cookies.

Author: done

## Cached artifacts

To faciliate efficient token acquisition while also maintaining a good UX, MSAL caches various artifacts resulting from its API calls. At a glance, these are:

- Persistent artifacts 
    - access tokens
    - id tokens
    - refresh tokens
    - accounts
- Temporary artifacts
    - request metadata (e.g. state, nonce, authority)
    - errors
    - interaction status
- Telemetry
    - previous failed request 
    - performance data

## Remarks

- MSAL Browser's cache schema is compatible with other MSALs. However, there is no cache sharing support at the moment. 
- We do not recommend apps having business logic dependent on direct use of entities in the cache. Instead, use the appropriate MSAL API when you need to acquire tokens or retrieve accounts.

> - Keys used to encrypt proof of possession (PoP) tokens are stored using the [IndexedDB API](https://developer.mozilla.org/docs/Web/API/IndexedDB_API). For more information, please refer to [access-token-proof-of-possession](./access-token-proof-of-possession.md#pop-key-management).

Reviewer 3: POP will use memory storage and IDB together.

Author: revised
